pkg/sql/audit_logging.go

```diff
@@ -122,7 +122,22 @@ func (p *planner) initializeReducedAuditConfig(ctx context.Context) {
 	p.reducedAuditConfig.Initialized = true
 
 	user := p.User()
-	userRoles, err := p.MemberOfWithAdminOption(ctx, user)
+	userRoles, err := func() (map[username.SQLUsername]bool, error) {
+		if p.Txn() != nil {
+			return p.MemberOfWithAdminOption(ctx, user)
+		} else {
+			// If there is no open transaction, then we need to create an internal
+			// one. This is the case for logging BEGIN statements, since the
+			// transaction is not opened until after BEGIN is logged.
+			var userRoles map[username.SQLUsername]bool
+			innerErr := p.ExecCfg().InternalDB.DescsTxn(ctx, func(ctx context.Context, txn descs.Txn) error {
+				var err error
+				userRoles, err = MemberOfWithAdminOption(ctx, p.ExecCfg(), txn, user)
+				return err
+			})
+			return userRoles, innerErr
```

Comment:
why can't the logic in the `if`/`else` clause be inside the `MemberOfWithAdminOption` method? This same logic is repeated above on lines 58-70.

Reply:
👋 thanks for asking! i think this change is lower-risk. for more context, the nil txn check inside of `MemberOfWithAdminOption` was added in https://github.com/cockroachdb/cockroach/pull/46246. in this PR, we want to handle the nil txn with `BEGIN` statements while still being careful/maintaining the invariant that the `MemberOfWithAdminOption` lookup needs a txn

so the main reason is: we want to make sure that we don't allow some undefined behavior if we just apply this "hey if theres a nil txn let's just create this internal txn" to the `MemberOfWithAdminOption` lookup everytime